ceph: update snapshot APIs from v1beta1 to v1 (#7711)
rook/rook: Documentation/ceph-upgrade.md

update to use the latest Ceph-CSI drivers. See the [CSI Version](#csi-version) section for more
details.

> Note: Refer to [ceph-csi-snapshot.md](ceph-csi-snapshot.md) for information on Snapshot API upgrade.

> 

Reviewer: nit: remove the extra line.

update to use the latest Ceph-CSI drivers. See the [CSI Version](#csi-version) section for more
details.

> > Note: Refer to [ceph-csi-snapshot.md](ceph-csi-snapshot.md) for information on Snapshot API upgrade.

Reviewer 2: ```suggestion
> Note: If using snapshots, refer to the [Upgrade Snapshot API guide](ceph-csi-snapshot.md#upgrade-snapshot-api).
```

ROOK_CSI_CEPH_IMAGE: "quay.io/cephcsi/cephcsi:v3.3.0"
ROOK_CSI_REGISTRAR_IMAGE: "k8s.gcr.io/sig-storage/csi-node-driver-registrar:v2.0.1"
ROOK_CSI_PROVISIONER_IMAGE: "k8s.gcr.io/sig-storage/csi-provisioner:v2.0.4"

> ROOK_CSI_SNAPSHOTTER_IMAGE: "k8s.gcr.io/sig-storage/csi-snapshotter:v4.0.0"

Reviewer 2: The snapshotter v4.0 is only compatible with K8s 1.20+? Seems like we need to have a different default depending on the K8s version where Rook is running.

Author: snapshotter siddecar v4.0 maintains backward compatibility https://github.com/kubernetes-csi/external-snapshotter/tree/master#snapshot-v1-apis, so it willl work with v1beta1 APIs snapshot controller and CRDs from k8s 1.17+.

Reviewer 2: Ok, so if someone installs Rook on K8s 1.16, what is the behavior? Will the snapshotter v4.0 be started, and in a failing state since it's not supported on alpha snapshots?

Author: It'll just throw the following warning like in other versions.
`E0423 11:59:56.144386       1 reflector.go:156] github.com/kubernetes-csi/external-snapshotter/pkg/client/informers/externalversions/factory.go:117: Failed to list *v1.VolumeSnapshotClass: the server could not find the requested resource (get volumesnapshotclasses.snapshot.storage.k8s.io)`

Users will need to downgrade snapshooter to <=v2.x to use alpha snapshot.

Reviewer 2: To avoid the error condition, how about we change the default image based on the K8s version? We're already checking the K8s version for the csi driver, for example [here](https://github.com/rook/rook/blob/a9641a9355e32eb56286f8b79cfd1c24b98c2438/pkg/operator/ceph/csi/spec.go#L351). We could use the v2.x snapshotter for the lesser version of K8s.

Author: This is the same warnings given in a k8s cluster with no snapshot CRDs and controller deployed. 

We can set the default snapshotter image to **v1.x*** for alpha snapshots but I think it's better to leave it up to the user to change the version which also anyway requires snapshot CRDs and setting features gates in <=1.16 . @Madhu-1 opinions?

Reviewer 3: IMO we should not worry about the Alpha version we can disable the snapshotter for Kubernetes 1.16 by default Enable it only if the user wants. For the version check instead of the kubernetes version check how about check for the snapshot CRD version? this check is only for default images if the user has provided some other custom image we should use it as it is.

ROOK_CSI_CEPH_IMAGE: "quay.io/cephcsi/cephcsi:v3.3.1"
ROOK_CSI_REGISTRAR_IMAGE: "k8s.gcr.io/sig-storage/csi-node-driver-registrar:v2.0.1"
ROOK_CSI_PROVISIONER_IMAGE: "k8s.gcr.io/sig-storage/csi-provisioner:v2.0.4"
ROOK_CSI_ATTACHER_IMAGE: "k8s.gcr.io/sig-storage/csi-attacher:v3.0.2"
ROOK_CSI_RESIZER_IMAGE: "k8s.gcr.io/sig-storage/csi-resizer:v1.0.1"
# if K8s <= v1.16, alpha snapshotter image is set as default
# ROOK_CSI_SNAPSHOTTER_IMAGE: "k8s.gcr.io/sig-storage/csi-snapshotter:v1.2.2"
# if K8s >= v1.17

> ROOK_CSI_SNAPSHOTTER_IMAGE: "k8s.gcr.io/sig-storage/csi-snapshotter:v4.0.0"

Reviewer 3: v4.x is only for Kubernetes 1.20+ https://kubernetes-csi.github.io/docs/external-snapshotter.html
